### PR TITLE
Bug 1168849 - Bookmark badge on awesome bar

### PR DESCRIPTION
**Pull request by fluffyemily**
https://bugzilla.mozilla.org/show_bug.cgi?id=1168849#add_comment
migrated and updated from a wesj patch that got bumped from 1.0 - https://github.com/mozilla/firefox-ios/pull/514

ensure that SQL is correct by adding spaces in all the right places
don't force unwrap optional in test assert - causes crash when test is failing

### CODE REVIEW

File: Storage/History.swift [hunk 1]
```diff
@@ -24,7 +24,7 @@ public protocol BrowserHistory {
     func removeSiteFromTopSites(site: Site) -> Success
 
     func getSitesByFrecencyWithLimit(limit: Int) -> Deferred<Maybe<Cursor<Site>>>
-    func getSitesByFrecencyWithLimit(limit: Int, whereURLContains filter: String) -> Deferred<Maybe<Cursor<Site>>>
+    func getSitesByFrecencyWithLimit(limit: Int, whereURLContains filter: String, includeBookmarked: Bool) -> Deferred<Maybe<Cursor<Site>>>
```
Comment (rnewman) [inline Storage/History.swift]: I think this name is misleading. We're not including bookmarked items; we're including whether or not the results were bookmarked. Maybe `includeBookmarkState`?

File: Client/Frontend/Browser/SearchViewController.swift [hunk 2]
```diff
@@ -499,6 +499,13 @@ extension SearchViewController {
                 if let cell = cell as? TwoLineTableViewCell {
                     cell.setLines(site.title, detailText: site.url)
                     cell.imageView?.setIcon(site.icon, withPlaceholder: self.defaultIcon)
+
+                    cell.rightBadge.image = nil
+                    if let bookmarked = site.bookmarked {
+                        if bookmarked {
```
Comment (rnewman) [inline Client/Frontend/Browser/SearchViewController.swift]:
```
if site.bookmarked ?? false {
```

File: Storage/SQL/SQLiteHistory.swift [hunk 3]
```diff
@@ -283,6 +283,10 @@ extension SQLiteHistory: BrowserHistory {
             site.latestVisit = Visit(date: latest, type: VisitType.Unknown)
         }
 
+        if let bookmarked = row["bookmarked"] as? Int {
```
Comment (rnewman) [inline Storage/SQL/SQLiteHistory.swift]: Perhaps add an optional variant to `extension SDRow.getBoolean` elsewhere in this file?

File: Storage/SQL/SQLiteHistory.swift [hunk 4]
```diff
@@ -320,24 +325,28 @@ extension SQLiteHistory: BrowserHistory {
...
-        "INNER JOIN \(TableDomains) ON \(TableDomains).id = \(TableHistory).domain_id " +
-        "INNER JOIN \(TableVisits) ON \(TableVisits).siteID = \(TableHistory).id " +
-        whereClause + " GROUP BY historyID"
+            "SELECT \(TableHistory).id AS historyID, \(TableHistory).url AS url, title, guid, domain_id, domain, " +
+            "COALESCE(max(case \(TableVisits).is_local when 1 then \(TableVisits).date else 0 end), 0) AS localVisitDate, " +
+            "COALESCE(max(case \(TableVisits).is_local when 0 then \(TableVisits).date else 0 end), 0) AS remoteVisitDate, " +
+            "COALESCE(count(\(TableVisits).is_local), 0) AS visitCount " +
+            (includeBookmarked ? ", count(\(TableBookmarks).id) AS bookmarked " : "") +
+            "FROM \(TableHistory) " +
+            (includeBookmarked ? "LEFT OUTER JOIN \(TableBookmarks) ON \(TableBookmarks).url IS \(TableHistory).url " : "") +
```
Comment (rnewman) [inline Storage/SQL/SQLiteHistory.swift]: I have a gentle preference for keeping the zero indent to buy more column width for SQL, and a moderate preference for not breaking blame — particularly not mixed in with a patch that makes other changes. If you're going to change spacing on lots of lines, do it in a separate commit so the real changes are easier to see.

Comment (rnewman) [inline Storage/SQL/SQLiteHistory.swift]: I have my doubts about joining history directly against bookmarks here, particularly using the URL. This is already a pretty expensive query.

The obvious long-term solution is to switch to a genuine places-style system with some shared indexed ID.

Other alternatives are to do a virtual join asynchronously after showing the list (which we should also do for icons), or to very simply… track bookmark state as a counter in each history item. Bookmarking is a rare event, adding history is a cheap place to do this query, etc. etc.

Can you do some analysis about the complexity change in the `EXPLAIN QUERY PLAN` here, and how this affects timing on moderate history DBs (~5000 history items) and moderate bookmark sets (~100)?

File: Storage/SQL/SQLiteHistory.swift [hunk 5]
```diff
@@ -411,11 +424,11 @@ extension SQLiteHistory: BrowserHistory {
         ", max(remoteVisitDate) AS remoteVisitDate" +
         ", sum(localVisitCount) AS localVisitCount" +
         ", sum(remoteVisitCount) AS remoteVisitCount" +
-        ", sum(frecency) AS frecencies" +
+        ", sum(frecency) AS frecencies\(bookedmarkedSQL)" +
```
Comment (rnewman) [inline Storage/SQL/SQLiteHistory.swift]: `+` on a new line instead.

File: Storage/SQL/SQLiteHistory.swift [hunk 6]
```diff
@@ -381,22 +391,25 @@ extension SQLiteHistory: BrowserHistory {
...
         whereClause + " GROUP BY historyID"
 
+        let bookedmarkedSQL = includeBookmarked ? ", bookmarked" : ""
         // Next: limit to only those that have been visited at all within the last six months.
         // (Don't do that in the innermost: we want to get the full count, even if some visits are older.)
         // Discard all but the 1000 most frecent.
         // Compute and return the frecency for all 1000 URLs.
         let frecenciedSQL =
-        "SELECT *, (\(localFrecencySQL) + \(remoteFrecencySQL)) AS frecency" +
+        "SELECT *, (\(localFrecencySQL) + \(remoteFrecencySQL)) AS frecency\(bookedmarkedSQL)" +
```
Comment (rnewman) [inline Storage/SQL/SQLiteHistory.swift]: `+` on a new line instead.

File: Storage/SQL/SQLiteHistory.swift [hunk 7]
```diff
@@ -424,9 +437,9 @@ extension SQLiteHistory: BrowserHistory {
             let sql = "SELECT" +
                       " historyID, url, title, guid, domain_id, domain" +
                       ", localVisitDate, remoteVisitDate, localVisitCount, remoteVisitCount" +
-                      ", iconID, iconURL, iconDate, iconType, iconWidth" +
-                      " FROM (\(historySQL)) LEFT OUTER JOIN " +
-                      "view_history_id_favicon ON historyID = view_history_id_favicon.id"
+                      ", iconID, iconURL, iconDate, iconType, iconWidth\(bookedmarkedSQL)" +
```
Comment (rnewman) [inline Storage/SQL/SQLiteHistory.swift]: `+` on a new line instead.

**Comment by fluffyemily** [comment]
@rnewman: I've fixed up those small things and then added a post query for fetching bookmarks if required. I felt that reorganizing bookmarks into a whole new storage structure was best left for some deeper thinking and a more important issue....

File: Storage/SQL/SQLiteHistory.swift [hunk 8]
```diff
@@ -349,18 +352,31 @@ extension SQLiteHistory: BrowserHistory {
                 "FROM (\(historySQL)) LEFT OUTER JOIN " +
                 "view_history_id_favicon ON historyID = view_history_id_favicon.id"
             let factory = SQLiteHistory.iconHistoryColumnFactory
-            return db.runQuery(sql, args: args, factory: factory)
+            return db.runQuery(sql, args: args, factory: factory) >>== { cursor in
+                if includeBookmarkState {
+                    // get list of tuples of ids & urls
```
Comment (rnewman) [inline Storage/SQL/SQLiteHistory.swift]: Nit:
```
// Augment the cursor with tuples of IDs and URLs.
```

File: Storage/SQL/SQLiteHistory.swift [hunk 9]
```diff
@@ -349,18 +352,31 @@ extension SQLiteHistory: BrowserHistory {
...
+                    return self.addBookmarkedState(toSites: cursor.asArray())
```
Comment (rnewman) [inline Storage/SQL/SQLiteHistory.swift]: Can we just take the cursor here?

That would also let us do

```
>== (includeBookmarkState ? self.addBookmarkedState : deferMaybe)
```

File: Storage/SQL/SQLiteHistory.swift [hunk 10]
```diff
@@ -349,18 +352,31 @@ extension SQLiteHistory: BrowserHistory {
...
+                return deferMaybe(cursor)
+            }
         }
 
         let factory = SQLiteHistory.basicHistoryColumnFactory
-        return db.runQuery(historySQL, args: args, factory: factory)
+        return db.runQuery(historySQL, args: args, factory: factory) >>== { cursor in
+            if includeBookmarkState {
+                // get list of tuples of ids & urls
+                return self.addBookmarkedState(toSites: cursor.asArray())
```
Comment (rnewman) [inline Storage/SQL/SQLiteHistory.swift]: Same comments.

File: Storage/SQL/SQLiteHistory.swift [hunk 11]
```diff
@@ -425,14 +441,48 @@ extension SQLiteHistory: BrowserHistory {
...
-                      " FROM (\(historySQL)) LEFT OUTER JOIN " +
-                      "view_history_id_favicon ON historyID = view_history_id_favicon.id"
+                      " FROM (\(historySQL)) LEFT OUTER JOIN" +
+                      " view_history_id_favicon ON historyID = view_history_id_favicon.id"
             let factory = SQLiteHistory.iconHistoryColumnFactory
-            return db.runQuery(sql, args: args, factory: factory)
+            return db.runQuery(sql, args: args, factory: factory) >>== { cursor in
+                if includeBookmarkState {
+                    // get list of tuples of ids & urls
+                    return self.addBookmarkedState(toSites: cursor.asArray())
```
Comment (rnewman) [inline Storage/SQL/SQLiteHistory.swift]: Same.

File: Storage/SQL/SQLiteHistory.swift [hunk 12]
```diff
@@ -425,14 +441,48 @@ extension SQLiteHistory: BrowserHistory {
...
+                return deferMaybe(cursor)
+            }
         }
 
         let factory = SQLiteHistory.basicHistoryColumnFactory
-        return db.runQuery(historySQL, args: args, factory: factory)
+        return db.runQuery(historySQL, args: args, factory: factory) >>== { cursor in
+            if includeBookmarkState {
+                // get list of tuples of ids & urls
+                return self.addBookmarkedState(toSites: cursor.asArray())
```
Comment (rnewman) [inline Storage/SQL/SQLiteHistory.swift]: Same.

File: Storage/SQL/SQLiteHistory.swift [hunk 13]
```diff
@@ -425,14 +441,48 @@ extension SQLiteHistory: BrowserHistory {
...
+            "GROUP BY \(TableHistory).id"
+        func BookmarkFactory(row: SDRow) -> Site {
+            // we should never fail this test as we will get a response containing only those rows that exist in the sites array, but optionals will be optionals
+            guard let siteID = row["historyID"] as? Int, let index = sites.indexOf({ $0.id == siteID })
+                else {
+                    log.error("Returning an empty site from Bookmarks. This shouldn't ever happen!");
+                    return Site(url: "", title: "")
+            }
+            let site = sites[index]
+            site.bookmarked = row.getBoolean("bookmarkCount")
```
Comment (rnewman) [inline Storage/SQL/SQLiteHistory.swift]: Ooh, relies on implementation detail of `getBoolean`! Let's make this explicitly `> 0`, please.

**Comment by rnewman** [comment]
I don't think this changes anything much *functionally* from the first version of your PR. It's much easier to read, though!

The cursor we return won't yield any results until a *second* query has run (indeed, we consume it immediately in order to decorate it), so this doesn't asynchronously badge rows — if the second query takes a while, we won't show anything at all until it's done. This is slightly slower than if we just wrapped the original query in another layer of joins (albeit easier to read and reason about)!

Neither does it precompute anything about bookmark state, which is the most efficient solution, but as you note would be much more invasive.

If you've done timing on, say, an iPhone 5 and can say "nope, that second query takes a millisecond, so it's not worth changing", great; just fix these things and land. If this round-trip is adding 100msec to an awesomebar completion for "www", then we probably need to go 'round again.

**Comment by thebnich** [comment]
@fluffyemily, looks like this didn't land because of some extra SQL stuff that was done. Can we split this into PRs as mentioned in https://bugzilla.mozilla.org/show_bug.cgi?id=1168849#c7, land the UI bits, and close this one out?
